File: snake.py
from collections import deque


class Snake:
    def __init__(
        self, start_col: int, start_row: int, grid_cols: int, grid_rows: int
    ) -> None:
        self._cols = grid_cols
        self._rows = grid_rows
        self._body: deque[tuple[int, int]] = deque(
            [
                (start_col, start_row),
                (start_col - 1, start_row),
                (start_col - 2, start_row),
            ]
        )
        self._direction: tuple[int, int] = (1, 0)
        self._pending: tuple[int, int] = (1, 0)
        self._grew: bool = False

    def set_direction(self, dx: int, dy: int) -> None:
        if (dx, dy) != (-self._direction[0], -self._direction[1]):
            self._pending = (dx, dy)

    def move(self) -> None:
        self._direction = self._pending
        hc, hr = self._body[0]
        new_head = (
            (hc + self._direction[0]) % self._cols,
            (hr + self._direction[1]) % self._rows,
        )
        self._body.appendleft(new_head)
        if not self._grew:
            self._body.pop()
        self._grew = False

    def grow(self) -> None:
        self._grew = True

    def head(self) -> tuple[int, int]:
        return self._body[0]

    def cell_in_front(self) -> tuple[int, int]:
        hc, hr = self._body[0]
        return (
            (hc + self._direction[0]) % self._cols,
            (hr + self._direction[1]) % self._rows,
        )

    def occupies(self, col: int, row: int) -> bool:
        return (col, row) in self._body

    def check_self_collision(self) -> bool:
        head = self._body[0]
        return any(seg == head for seg in list(self._body)[1:])

    @property
    def length(self) -> int:
        return len(self._body)

    def get_body_cells(self) -> list[tuple[int, int]]:
        return list(self._body)

File: src/game/snake.ts
import { mod } from './grid';
import type { Cell } from './types';

export class Snake {
  private readonly body: Cell[];
  private dc: number;
  private dr: number;
  private pendingGrowth: number;
  readonly cols: number;
  readonly rows: number;

  constructor(startCol: number, startRow: number, gridCols: number, gridRows: number) {
    this.cols = gridCols;
    this.rows = gridRows;
    this.dc = 1;
    this.dr = 0;
    this.body = [{ col: startCol, row: startRow }];
    this.pendingGrowth = 0;
  }

  setDirection(dc: number, dr: number): void {
    if (dc !== 0 && dc === -this.dc) return;
    if (dr !== 0 && dr === -this.dr) return;
    this.dc = dc;
    this.dr = dr;
  }

  move(): { didWrap: boolean } {
    // biome-ignore lint/style/noNonNullAssertion: body always has ≥1 element
    const head = this.body[0]!;
    const rawCol = head.col + this.dc;
    const rawRow = head.row + this.dr;
    const newCol = mod(rawCol, this.cols);
    const newRow = mod(rawRow, this.rows);
    const didWrap = newCol !== rawCol || newRow !== rawRow;

    this.body.unshift({ col: newCol, row: newRow });

    if (this.pendingGrowth > 0) {
      this.pendingGrowth--;
    } else {
      this.body.pop();
    }

    return { didWrap };
  }

  grow(): void {
    this.pendingGrowth++;
  }

  head(): Cell {
    // biome-ignore lint/style/noNonNullAssertion: body always has ≥1 element
    return this.body[0]!;
  }

  get length(): number {
    return this.body.length;
  }

  getBodyCells(): Cell[] {
    return [...this.body];
  }

  cellInFront(): Cell {
    // biome-ignore lint/style/noNonNullAssertion: body always has ≥1 element
    const h = this.body[0]!;
    return {
      col: mod(h.col + this.dc, this.cols),
      row: mod(h.row + this.dr, this.rows),
    };
  }

  checkSelfCollision(): boolean {
    // biome-ignore lint/style/noNonNullAssertion: body always has ≥1 element
    const h = this.body[0]!;
    return this.body.slice(1).some(c => c.col === h.col && c.row === h.row);
  }
}
